### DISCUSSION TITLE
How to return a Query in a (mutation) response

### DISCUSSION THREAD

**Discussion by magnayn-mns** [graphql-java/graphql-java#3193] [Q&A]
I'm trying to implement the "query object in a response from a mutation" pattern, described here:
https://graphql-rules.com/rules/mutation-payload-query

(Our stack includes graphql-java-federation)

I can't seem to get this to work; parsing the schema fails if I define Query.

I have been trying to modify the schema as I'd read that having a field which was a Type Reference might work, but I can't seem to modify a type this way (I end up with 2 definitions which obviously goes wrong).

Is there a known way to do this?

```java
final GraphQLSchema.Builder newSchema = GraphQLSchema.newSchema(graphQLSchema);
    GraphQLTypeReference qtr = GraphQLTypeReference.typeRef("Query");

    GraphQLObjectType response =
        (GraphQLObjectType) graphQLSchema.getType("Price_VoucherApplySuccess");

    GraphQLObjectType newResponse =
        response.transform(
            rr -> {
              
              rr.field(new GraphQLFieldDefinition.Builder().name("query").type(qtr));

              rr.build();
            });


    newSchema.additionalType(newResponse);
```

**Comment by bbakerman** [comment]
I dont think the "schema.transform" method will do what you want.  I cant see the state of the schema as it enters `GraphQLSchema.newSchema(graphQLSchema)` nor can I see the errors coming back to you.  If you could list ouyt this it might make it easier to debug.

BuT I think you want to use another mechanism called `SchemaTransformer` to make such a change.  It uses a visitor + "modification" mechanism to re-build schemas.

See https://www.graphql-java.com/documentation/schema#changing-schema

**Reply by bbakerman** [reply]
ps... in general your approach of adding a new field that points (type reference) back to the Query type seems right however give the SchemaTransformer approach a try
